File: src/main/java/org/swordess/common/vitool/cmd/JasyptCommands.java
package org.swordess.common.vitool.cmd;

import org.jasypt.encryption.StringEncryptor;
import org.jasypt.encryption.pbe.PooledPBEStringEncryptor;
import org.jasypt.encryption.pbe.config.SimpleStringPBEConfig;
import org.springframework.shell.Availability;
import org.springframework.shell.standard.ShellComponent;
import org.springframework.shell.standard.ShellMethod;
import org.springframework.shell.standard.ShellMethodAvailability;
import org.springframework.util.StringUtils;

@ShellComponent
public class JasyptCommands {

    private String password;

    @ShellMethod("Set the jasypt encryptor password")
    public void jasyptSetPassword(String value) {
        password = value;
    }

    @ShellMethod("Encrypt the given input string")
    public String jasyptEncrypt(String input) {
        return getDefaultEncryptor().encrypt(input);
    }

    @ShellMethod("Decrypt the given (encrypted) input string")
    public String jasyptDecrypt(String encryptedInput) {
        return getDefaultEncryptor().decrypt(encryptedInput);
    }

    @ShellMethodAvailability({ "jasyptEncrypt", "jasyptDecrypt" })
    public Availability availabilityCheck() {
        return !StringUtils.isEmpty(password)
                ? Availability.available()
                : Availability.unavailable("the jasypt encryptor password has not been set yet");
    }

    private StringEncryptor getDefaultEncryptor() {
        // The following config is originated from the maven artifact
        //   com.github.ulisesbocchio:jasypt-spring-boot:2.1.1
        //     com.ulisesbocchio.jasyptspringboot.encryptor.DefaultLazyEncryptor#createPBEDefault
        SimpleStringPBEConfig config = new SimpleStringPBEConfig();
        config.setPassword(password);
        config.setAlgorithm("PBEWithMD5AndDES");
        config.setKeyObtentionIterations(1000);
        config.setPoolSize(1);
        config.setSaltGeneratorClassName("org.jasypt.salt.RandomSaltGenerator");
        config.setIvGeneratorClassName("org.jasypt.salt.NoOpIVGenerator");
        config.setStringOutputType("base64");

        PooledPBEStringEncryptor encryptor = new PooledPBEStringEncryptor();
        encryptor.setConfig(config);
        return encryptor;
    }

}

File: src/main/kotlin/org/swordess/common/vitool/cmd/JasyptCommands.kt
/*
 * Copyright (c) 2019-2022 Swordess
 *
 * Distributed under MIT license.
 * See file LICENSE for detail or copy at https://opensource.org/licenses/MIT
 */

package org.swordess.common.vitool.cmd

import org.jasypt.encryption.StringEncryptor
import org.jasypt.encryption.pbe.PooledPBEStringEncryptor
import org.jasypt.encryption.pbe.config.SimpleStringPBEConfig
import org.springframework.shell.Availability
import org.springframework.shell.standard.ShellComponent
import org.springframework.shell.standard.ShellMethod
import org.springframework.shell.standard.ShellMethodAvailability

@ShellComponent
class JasyptCommands {

    private var password: String? = null

    @ShellMethod(key = ["jasypt set-password"], value = "Set the jasypt encryptor password.")
    fun jasyptSetPassword(value: String) {
        password = value
    }

    @ShellMethod(key = ["jasypt encrypt"], value = "Encrypt the given input string.")
    fun jasyptEncrypt(input: String): String = defaultEncryptor().encrypt(input)

    @ShellMethod(key = ["jasypt decrypt"], value = "Decrypt the given (encrypted) input string.")
    fun jasyptDecrypt(encryptedInput: String): String = defaultEncryptor().decrypt(encryptedInput)

    @ShellMethodAvailability("jasypt encrypt", "jasypt decrypt")
    fun availabilityCheck(): Availability =
        if (password?.isNotBlank() == true) Availability.available() else Availability.unavailable("the jasypt encryptor password has not been set yet")

    private fun defaultEncryptor(): StringEncryptor {
        // The following config is originated from the maven artifact
        //   com.github.ulisesbocchio:jasypt-spring-boot:2.1.1
        //     com.ulisesbocchio.jasyptspringboot.encryptor.DefaultLazyEncryptor#createPBEDefault
        val config = SimpleStringPBEConfig().apply {
            this.password = this@JasyptCommands.password
            algorithm = "PBEWithMD5AndDES"
            keyObtentionIterations = 1000
            poolSize = 1
            setSaltGeneratorClassName("org.jasypt.salt.RandomSaltGenerator")
            setIvGeneratorClassName("org.jasypt.salt.NoOpIVGenerator")
            stringOutputType = "base64"
        }

        return PooledPBEStringEncryptor().apply {
            setConfig(config)
        }
    }

}